diesel/src/expression/not.rs

```diff
@@ -25,7 +25,7 @@ use crate::sql_types::{Bool, Nullable};
 /// ```
 pub fn not<T>(expr: T) -> not<T>
 where
-    T: AsExpression<Nullable<Bool>>,
+    T: AsExpression<Bool>,
```

Comment:
I'm not sure if that should only accept non nullable values. I think accepting both `Bool` and `Nullable<Bool>` here is more meaningful. On the other hand that could potentially lead to situations where rustc cannot infer the correct type for `ST` from `AsExpression` here.

Reply:
Hmm that looks like kind of the same problem as `and`/`or` no? Where if we enforce that it takes an `AsExpression<Nullable<Bool>>` we'll never be able to make it return a non-nullable bool?

Additionally, not all expressions of SqlType `Bool` implement `AsExpression<Nullable<Bool>>` anymore, because otherwise this creates type inference issues with `and`/`or`: https://github.com/diesel-rs/diesel/pull/2597#discussion_r546080041
Hence the following does not compile anymore if we do this: `not(name.eq("Sean")` (from tests)

Finally, I believe the behavior of `not` on `Nullable<Bool>` is a bit of a trap in that `NULL` values result in `NULL`, so they do not change "truthiness" with regards to e.g. a `filter`, which may lead to nasty bugs in case [this kind of confusion](https://github.com/diesel-rs/diesel/pull/2597#issuecomment-748494079) is made and the compiler does not prevent it.

So overall I believe we should keep this behavior where we only accept non-nullable values here.

Reply:
> I think accepting both Bool and Nullable<Bool>

I had encountered a situation where I want to store the same. Is there a way we can specify that?

Reply:
We must accept both variants (`Bool` and `Nullable<Bool>`) here otherwise it's not possible to write queries like `not(foo.and(bar))` or even `not(foo.eq(bar))` anymore in cases one of the arguments is `Nullable<Bool>`

Reply:
> I had encountered a situation where I want to store the same

Not sure what you mean by "the same".

> We must accept both variants (`Bool` and `Nullable<Bool>`) here otherwise it's not possible to write queries like `not(foo.and(bar))` or even `not(foo.eq(bar))` anymore in cases one of the arguments is `Nullable<Bool>`

So I thought it maybe wasn't a good idea to have `not` support this implicitly, as this may easily lead to nasty bugs:

> I believe the behavior of `not` on `Nullable<Bool>` is a bit of a trap in that `NULL` values result in `NULL`, so they do not change "truthiness" with regards to e.g. a `filter`, which may lead to nasty bugs in case [this kind of confusion](https://github.com/diesel-rs/diesel/pull/2597#issuecomment-748494079) is made and the compiler does not prevent it.

Example:
```rust
table.filter(not(foo.eq(bar)))
```
Where `foo` = `true` and `bar` = `NULL`, the lines won't show, because `not` won't change the "truthiness" of the expression, which is a bit counter-intuitive.

Also, our codebase is rather large and we have zero such patterns.

Given that, do you still believe it's important for `not` to support this? (and we couldn't be happy with a separate `nullable_not`?)

Reply:
I still believe that `not` should support that as this is just the behavior of the underlying SQL operator. I feel that we had this discussion already about a few operators (for example `eq` in another context). Every time the final answer was: We follow the semantic of the underlying SQL operator as we do not try to hide complexity at this level of abstraction. Hiding this complexity can be done in a crate build on top of diesel, providing it owns set of operators that is not based on SQL in my opinion.

Reply:
I agree. Done in https://github.com/diesel-rs/diesel/pull/2597/commits/df8b5e59ef232cd2d517f2e764caab207b39088b.